### PR TITLE
fix (worker): upload with fallback on CDS API

### PR DESCRIPTION
**Pull request by yesnault**
Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

**Comment by yesnault** [comment]
this will avoid error as:

"Error while uploading artifact 'Error_Report.html': error on io.Copy(part, fileReopen), err: read c:\...\Error_Report.html: file already closed 
End of step /Artifact Upload-4 [Fail] with reason: Error while uploading artifact 'Error_Report.html': error on io.Copy(part, fileReopen), err: read c:\...\Error_Report.html: file already closed
"

**Comment by fsamin** [comment]
The message error `'Error_Report.html': error on io.Copy(part, fileReopen), err: read c:...\Error_Report.html: file already closed` just said that the reader (fileReopen) is over on the second attempt.

Don't need to fallback on API  to upload, we just had to read the buffer once and reopen a new reader on this buffer.

Streaming artifact through CDS API have a massive impact on memory consumption; we have to avoid it.

File: sdk/artifact.go [hunk 1]
```diff
@@ -220,30 +220,29 @@ func UploadArtifact(project string, pipeline string, application string, tag str
...
 	if errst != nil {
-		return false, 0, errst
+		return false, 0, fmt.Errorf("error on fileForMD5.Stat(), errst: %s", errst)
 	}
 
 	//Compute md5sum
 	hash := md5.New()
 	if _, errcopy := io.Copy(hash, fileForMD5); errcopy != nil {
-		return false, 0, errcopy
+		return false, 0, fmt.Errorf("error on io.Copy, errcopy:%s", errcopy)
```
Comment (fsamin) [inline sdk/artifact.go]: "unable to read file content (%v)", errcopy

File: sdk/artifact.go [hunk 2]
```diff
@@ -220,30 +220,29 @@ func UploadArtifact(project string, pipeline string, application string, tag str
...
 	hashInBytes := hash.Sum(nil)[:16]
 	md5sumStr := hex.EncodeToString(hashInBytes)
 	fileForMD5.Close()
 
 	//Reopen the file because we already read it for md5
-	fileReopen, erro := os.Open(filePath)
+	fileContent, erro := ioutil.ReadFile(filePath)
 	if erro != nil {
-		return false, 0, erro
+		return false, 0, fmt.Errorf("error os.Open(filePath), erro:%s", erro)
```
Comment (fsamin) [inline sdk/artifact.go]: "unable to read file %s (%v)", fileContent , erro

File: sdk/artifact.go [hunk 3]
```diff
@@ -252,22 +251,23 @@ func UploadArtifact(project string, pipeline string, application string, tag str
 		_ = json.Unmarshal(bodyRes, store)
 
 		if store.TemporaryURLSupported {
-			tempURL, dur, err := uploadArtifactWithTempURL(project, pipeline, application, env, tag, buildNumber, name, fileReopen, stat, md5sumStr)
+			tempURL, dur, err := uploadArtifactWithTempURL(project, pipeline, application, env, tag, buildNumber, name, fileContent, stat, md5sumStr)
 			if err == nil {
-				return tempURL, dur, err
+				return tempURL, dur, fmt.Errorf("error on uploadArtifactWithTempURL err:%s", err)
```
Comment (fsamin) [inline sdk/artifact.go]: "unable to upload to objectstore (%v)", err

File: sdk/artifact.go [hunk 4]
```diff
@@ -252,22 +251,23 @@ func UploadArtifact(project string, pipeline string, application string, tag str
...
 		}
 	}
 
+	// if we are here, upload with tempURL didn't work. Fallback to download with CDS API.
 	body := new(bytes.Buffer)
 	writer := multipart.NewWriter(body)
 	part, errc := writer.CreateFormFile(name, filepath.Base(filePath))
 	if errc != nil {
-		return false, 0, errc
+		return false, 0, fmt.Errorf("error on writer.CreateFormFile() errc:%s", errc)
```
Comment (fsamin) [inline sdk/artifact.go]: "unable to create multipart form file (%v)", errc

File: sdk/artifact.go [hunk 5]
```diff
@@ -252,22 +251,23 @@ func UploadArtifact(project string, pipeline string, application string, tag str
...
 	}
 
-	if _, err := io.Copy(part, fileReopen); err != nil {
-		return false, 0, err
+	if _, err := io.Copy(part, bytes.NewReader(fileContent)); err != nil {
+		return false, 0, fmt.Errorf("error on io.Copy(part, bytes.NewReader(fileContent)), err: %s", err)
```
Comment (fsamin) [inline sdk/artifact.go]: "unable to read file content (%v)", err

File: sdk/artifact.go [hunk 6]
```diff
@@ -276,7 +276,7 @@ func UploadArtifact(project string, pipeline string, application string, tag str
 	writer.WriteField("md5sum", md5sumStr)
 
 	if err := writer.Close(); err != nil {
-		return false, 0, err
+		return false, 0, fmt.Errorf("error on write.Close(), err:%s", err)
```
Comment (fsamin) [inline sdk/artifact.go]: "unable to close multipart form writer(%v)", err

**Pull request merged**